Fix issue with hoisting and copy prop interaction

BruceForstall opened the pull request:
After hoisting creates a tree copy, it morphs it. That morph might lose the value numbers on LCL_VAR uses, but leave the SSA numbers. Copy prop was assuming that such a use had a VN. Instead, check this dynamically instead of asserting.

Fixes #84619

BruceForstall committed:
Fix issue with hoisting and copy prop interaction

After hoisting creates a tree copy, it morphs it. That morph might
lose the value numbers on LCL_VAR uses, but leave the SSA numbers.
Copy prop was assuming that such a use had a VN. Instead, check this
dynamically instead of asserting.

Fixes #84619

BruceForstall commented:
/azp run Fuzzlyn

BruceForstall commented:
[No asm diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=255257&view=ms.vss-build-web.run-extensions-tab)

Slight TP cost

BruceForstall commented:
@AndyAyersMS PTAL
cc @dotnet/jit-contrib

AndyAyersMS commented:
> After hoisting creates a tree copy, it morphs it. That morph might lose the value numbers on LCL_VAR uses, but leave the SSA numbers. Copy prop was assuming that such a use had a VN. Instead, check this dynamically instead of asserting.
> 
> Fixes #84619

I hate to open a can of worms here, but *why* does hoisting call morph? Seems like maybe it shouldn't.

BruceForstall commented:
> I hate to open a can of worms here, but _why_ does hoisting call morph?

Disabling morphing only causes one family of diffs.

Another question, though: why does morphing lose VNs on LCL_VARs?

SingleAccretion commented:
The case from #84619 is `optNarrowTree`'s work, and it cannot (easily) preserve VNs on trees that it transforms (although it has some logic that tries). What we've been doing with such transformation is putting them under `if (fgGlobalMorph)`. This may be expensive CQ-wise in this case.

AndyAyersMS commented:
> > I hate to open a can of worms here, but _why_ does hoisting call morph?
> 
> Disabling morphing only causes one family of diffs.

I was wondering if you'd actually see improvements, since hoisting relies on CSE and perhaps morphing the hoisted copy would break CSE (either by changing tree shape so it no longer matches the in-loop tree or perhaps by losing the `GTF_MAKE_CSE` flag) and hence messes up hoisting.

Oddly we don't check for `GTF_MAKE_CSE` in CSE proper, we just use it to keep morph and other things from deleting the thing we just hoisted.

> Another question, though: why does morphing lose VNs on LCL_VARs?

Could be there's a missing surgical VN update that could be fixed, you'd have to dig in.

AndyAyersMS commented:
@BruceForstall you're not waiting on me here, are you?

BruceForstall commented:
> @BruceForstall you're not waiting on me here, are you?

No. But the questions asked require investigations, which I've prioritized very low compared to other work (even though it would be nice to get Fuzzlyn cleaner, but it has good baselining anyway)

BruceForstall committed:
Further optimize cast in fgOptimizeCastOnStore()

If we optimize the cast, call `fgOptimizeCast` to see if it can be
optimized further.

BruceForstall committed:
Add regression test

BruceForstall commented:
@AndyAyersMS I updated this. I found the case that was creating a CAST that wasn't getting fully morphed/optimized, in `fgOptimizeCastOnStore`, and added a call to `fgOptimizeCast`. With this, the hoisted code that gets morphed doesn't actually optimize anything, and so the assert is avoided. (It turns out it doesn't create a CSE in this test case, either, but I didn't look into why.) I propose to leave the dynamic check in `optBlockCopyProp` for safety. Finally, if I remove morphing from hoisting there are no diffs now (at least on win-x64). Perhaps removing that can be done as a follow-up.

I also added a unit test.

There are a few [diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=274714&view=ms.vss-build-web.run-extensions-tab), with the additional call to `fgOptimizeCast` causing a few more cases of narrowing to kick in.

There is a bit of TP hit.

SingleAccretion commented on src/coreclr/jit/copyprop.cpp:
```diff
@@ -389,7 +391,8 @@ bool Compiler::optBlockCopyProp(BasicBlock* block, LclNumToLiveDefsMap* curSsaNa
                 optCopyPropPushDef(tree, lclDefNode, curSsaName);
             }
-            else if (tree->OperIs(GT_LCL_VAR, GT_LCL_FLD) && tree->AsLclVarCommon()->HasSsaName())
+            else if (tree->OperIs(GT_LCL_VAR, GT_LCL_FLD) && tree->AsLclVarCommon()->HasSsaName() &&
+                     tree->gtVNPair.BothDefined())
```
I would recommend removing this check. `optCopyProp` doesn't actually depends on the VNs being present for trees it examines, so this is effectively the same as removing the assert.

BruceForstall replied:
`optCopyProp` does `assert(tree->gtVNPair.BothDefined())` which is why I added this.

But it's true that `optCopyProp` does not actually use the `tree->gtVNPair` : it looks up the VN in the `lvPerSsaData` for the SSA def of the local. Presumably this assert is a partial expression of the expectation that if a local has an SSA number, then its SSA def should have the same VN as the SSA use (I believe that should be true?). It would therefore seem odd to ignore the fact that a lclvar use has no VN but its SSA does, and just use the def. That seems to indicate something is odd/wrong, perhaps due to morph, like what happened here.

Of course, I could remove the assert from `optCopyProp`, too.

SingleAccretion replied:
> Presumably this assert is a partial expression of the expectation that if a local has an SSA number, then its SSA def should have the same VN as the SSA use (I believe that should be true?).

Here is how this situation came to be: `optCopyProp` used to use the tree VN, but was changed to use the definition VN in #65250, with the main motivation being `LCL_FLD` handling. I left the tree VN assert as it seemed valuable on its own (indeed, as you say, to catch missing VNs - we are generally sorely lacking in validating this aspect of morph).

BruceForstall replied:
So it seems like I could either:
1. Remove both the `tree->gtVNPair.BothDefined()` check here and `assert(tree->gtVNPair.BothDefined())` in `optCopyProp` (so `optCopyProp` will no longer check for "missing" VNs), or
2. Leave it as I've written it -- with more testing/checking than maybe should be required, or
3. Something else? (e.g., asserting for and fixing "missing VNs" -- but that seems out of scope for this fix)

Do you have a preference?

SingleAccretion replied:
My preference would be to check for missing VNs and fix them (leave the assert without the check, or even migrate it to the general loop in `optBlockCopyProp`), but barring that removing the assert (`1`) seems better.

AndyAyersMS replied:
I think the SSA checker should also be able to spot missing VNs, though not perhaps in as timely a manner.

BruceForstall replied:
> I think the SSA checker should also be able to spot missing VNs, though not perhaps in as timely a manner.

Well, `fgDebugCheckSsa()` already has this comment: `* Verify VNs on uses match the VN on the def` :-)

BruceForstall committed:
Feedback

BruceForstall commented:
Ok, I've removed the assert in `optCopyProp`.

BruceForstall commented:
ping
